examples/monad.kl

```diff
@@ -0,0 +1,231 @@
+
+(datatype (Maybe A)
+  (nothing)
+  (just A))
+
+(define maybe-functor
+  (functor
+    (lambda (f x)
+      (case x
+        [(nothing) (nothing)]
+        [(just x) (just (f x))]))))
+
+(define macro-functor
+  (functor
+    (lambda (f x)
+      (>>= x (lambda (y) (pure (f y)))))))
+
+(define fmap
+  (lambda (inst f x)
+    (case inst
+      [(functor op) (op f x)])))
+
+(example fmap)
+
+(datatype (Applicative F A B)
+  (applicative (Functor F A B)
+               (-> A (F A))
+               (-> (F (-> A B)) (F A) (F B))))
+
+
+(define macro-applicative
+  (applicative
+    macro-functor
+    (lambda (x) (pure x))
+    (lambda (fun arg)
+      (>>= fun (lambda (g) (>>= arg (lambda (x) (pure (g x)))))))))
+
+(example macro-applicative)
+
+(define maybe-applicative
+  (applicative
+    maybe-functor
+    (lambda (x) (just x))
+    (lambda (fun arg)
+      (case fun
+        ((nothing) (nothing))
+        ((just f)
+         (case arg
+           ((nothing) (nothing))
+           ((just x) (just (f x)))))))))
+
+(define return
+  (lambda (inst x)
+    (case inst
+      [(applicative fun p ap) (p x)])))
+
+(example return)
+
+(define ap
+  (lambda (inst f arg)
+    (case inst
+      [(applicative fun p ap) (ap f arg)])))
+
+(example ap)
+
+(define-macros
+  ([idiom
+    (lambda (stx)
+      (syntax-case stx
+        [(cons _ inst-and-args)
+         (syntax-case inst-and-args
+           [(cons inst args)
+            (syntax-case args
+              [(cons fun more)
+               (pure
+                (quasiquote/loc stx
+                  (let (inst ,inst)
+                    (idiom-helper inst (return inst ,fun) ,more))))])])]))]
+   [idiom-helper
+    (lambda (stx)
+      (syntax-case stx
+        [(list (_ inst fun args))
+         (syntax-case args
+           [() (pure fun)]
+           [(cons x xs)
+            (pure
+             (quasiquote/loc stx
+               (idiom-helper ,inst (ap ,inst ,fun ,x) ,xs)))])]))]))
+
+(example (idiom maybe-applicative string-append (just "applicative") (just " notation")))
+(example (idiom maybe-applicative string-append (just "applicative") (nothing)))
+
+(define-macros
+  ([define-idiom
+    (lambda (stx)
+      (syntax-case stx
+       [(list (_ kw inst))
+        (pure
+         (quasiquote/loc stx
+           (group
+            (define inst ,inst)
+            (define-macros
+              ([,kw
+                (lambda (stx)
+                  (syntax-case stx
+                    [(cons _ args)
+                     (pure (cons-list-syntax 'idiom (cons-list-syntax 'inst args stx) stx))]))])))))]))]))
+
+(define-idiom perhaps maybe-applicative)
+
+(example (perhaps string-append (just "applicative") (just " notation")))
+(example (perhaps string-append (just "applicative") (nothing)))
+
+
+(datatype (Monad M A B)
+  (monad (Applicative M A B)
+         (-> (M A) (-> A (M B))
+             (M B))))
+
+(define macro-monad
+  (monad macro-applicative (lambda (m f) (>>= m f))))
```

Comment:
I see that the `lambda` is necessary because `bindMacro` is a macro, not a function. I don't remember why we defined it this way, is there a good reason?